c10/cuda/CUDAStream.cpp

```diff
@@ -225,59 +233,60 @@ cudaStream_t CUDAStream::stream() const {
-          " official API like c10::cuda::getStreamFromPool() to get a new stream.");
-      return nullptr;
-    case StreamIdType::LOW:
-      return low_priority_streams[device_index][si];
-    case StreamIdType::HIGH:
-      return high_priority_streams[device_index][si];
-    case StreamIdType::EXT:
-      return reinterpret_cast<cudaStream_t>(stream_id);
-    default:
-      TORCH_INTERNAL_ASSERT(
-          0,
-          "Unrecognized stream ",
-          stream_,
-          " (I didn't recognize the stream type, ",
-          st,
-          ")");
+  if (st.isDefault()) {
+    TORCH_INTERNAL_ASSERT(
+        si == 0,
+        "Unrecognized stream ",
+        stream_,
+        " (I think this should be the default stream, but I got a non-zero index ",
+        si,
+        ").",
+        " Did you manufacture the StreamId yourself?  Don't do that; use the",
+        " official API like c10::cuda::getStreamFromPool() to get a new stream.");
+    return nullptr;
+  } else if (st.isExt()) {
+    return reinterpret_cast<cudaStream_t>(stream_id);
+  } else {
+    auto streamType = st.getStreamType();
+    TORCH_INTERNAL_ASSERT(
+        streamType >= 1 && streamType <= max_stream_priorities,
+        "Unrecognized stream ",
+        stream_,
+        " (I didn't recognize the stream type, ",
+        streamType,
+        ")");
+    return streams[st.getStreamType() - 1][device_index][si];
   }
 }
 
 // Returns a stream from the requested pool
 // Note: when called the first time on a device, this will create the
 // stream pools for that device.
-CUDAStream getStreamFromPool(
-    const bool isHighPriority,
-    DeviceIndex device_index) {
+CUDAStream getStreamFromPool(const int priority, DeviceIndex device_index) {
   initCUDAStreamsOnce();
   if (device_index == -1) {
     device_index = current_device();
     c10::cuda::SetTargetDevice();
   }
+  TORCH_CHECK(
+      priority <= 0,
+      "Expected cuda stream priority to be less than or equal to 0, got ",
+      priority);
   check_gpu(device_index);
-
   // Initializes the stream pools (once)
   c10::call_once(
       device_flags[device_index], initDeviceStreamState, device_index);
+  auto pri_idx = -priority;
+  pri_idx =
+      std::min(pri_idx, max_stream_priorities - 1); // pri_idx is zero-based
+  const auto idx = get_idx(priority_counters[pri_idx][device_index]);
+  StreamIdType id_type = StreamIdType(pri_idx + 1);
+  return CUDAStreamForId(device_index, makeStreamId(id_type, idx));
+}
 
-  if (isHighPriority) {
-    const auto idx = get_idx(high_priority_counters[device_index]);
-    return CUDAStreamForId(device_index, makeStreamId(StreamIdType::HIGH, idx));
-  }
-
-  const auto idx = get_idx(low_priority_counters[device_index]);
-  return CUDAStreamForId(device_index, makeStreamId(StreamIdType::LOW, idx));
+CUDAStream getStreamFromPool(const bool isHighPriority, DeviceIndex device) {
+  int priority = isHighPriority ? -max_stream_priorities + 1 : 0;
```

Comment:
This is the "use the highest priority" priority implementation

Reply:
Yes, this is the highest priority, @chillee said that that's what he wants